src/node_file.cc

```diff
@@ -3417,6 +3419,214 @@ static void CpSyncOverrideFile(const FunctionCallbackInfo<Value>& args) {
+                THROW_ERR_FS_CP_EINVAL(env,
+                                       message.c_str(),
+                                       target.c_str(),
+                                       current_target.c_str());
+                return false;
+              }
+
+              // Prevent copy if src is a subdir of dest since unlinking
+              // dest in this case would result in removing src contents
+              // and therefore a broken symlink would be created.
+              if (std::filesystem::is_directory(dest_file) &&
+                  isInsideDir(current_target, target)) {
+                std::string message = "cannot overwrite %s with %s";
+                THROW_ERR_FS_CP_SYMLINK_TO_SUBDIRECTORY(env,
+                                                        message.c_str(),
+                                                        current_target.c_str(),
+                                                        target.c_str());
+                return false;
+              }
+
+              // symlinks get overridden by cp even if force: false, this is
+              // being applied here for backward compatibility, but is it
+              // correct? or is it a bug?
+              std::filesystem::remove(dest_file, error);
+              if (error) {
+                env->ThrowStdErrException(error, "cp", dest.c_str());
+                return false;
+              }
+            } else if (std::filesystem::is_regular_file(dest_file)) {
+              if (!dereference || (!force && error_on_exist)) {
+                env->ThrowStdErrException(
+                    std::make_error_code(std::errc::file_exists),
+                    "cp",
+                    dest_file.c_str());
+                return false;
+              }
+            }
+          }
+
+          if (dir_entry.is_directory()) {
+            std::filesystem::create_directory_symlink(
+                target_absolute, dest_file, error);
+          } else {
+            std::filesystem::create_symlink(target_absolute, dest_file, error);
+          }
+          if (error) {
+            env->ThrowStdErrException(error, "cp", dest.c_str());
+            return false;
+          }
+        }
+      } else if (dir_entry.is_directory()) {
+        auto src_entry_path = std::filesystem::relative(dir_entry.path(), src);
+        auto filename = dir_entry.path().filename();
+        std::filesystem::create_directory(dest_file);
+        auto success = copy_dir_contents(src / filename, dest_file);
+        if (!success) {
+          return false;
+        }
+      } else if (dir_entry.is_regular_file()) {
+        auto src_entry_path = std::filesystem::relative(dir_entry.path(), src);
+        auto dest_entry_path = dest / src_entry_path;
+        std::filesystem::copy_file(
+            dir_entry.path(), dest_entry_path, file_copy_opts, error);
+        if (error) {
+          if (error.value() == EEXIST) {
+            THROW_ERR_FS_CP_EEXIST(
```

Comment:
I was thinking about this throw... and I think that it could be nicer to just throw the EEXIST error from the C++ layer, catch it in the js layer and then throw an `ERR_FS_CP_EEXIST` error from there (using the [appropriate class](https://github.com/nodejs/node/blob/d89657c29e69043289ae0f75d87cca634d396bff/lib/internal/errors.js#L1211)).

I did not however go with that solution as I think that catching and re-throwing in the js layer interferes with userland debuggers that can catch those as part of options such as `Pause on caught exceptions`, please let me know if you think I'm wrong and that throwing the error from the js layer could be a better choice here :slightly_smiling_face: